Add bitstream replacement page and submission modal

gingyx opened the pull request:
## References
- Fixes #4104
- REST PR: DSpace/DSpace#10554

## Description
This PR provides the angular side for the REST PR mentioned above. It introduces a bitstream replace page and provides a bitstream replace modal during item submission.

## Instructions for Reviewers
List of changes in this PR:

1. Created `ReplaceBitstreamPageComponent` in which users can upload and replace an existing bitstream.
2. Added a new path `../bitstreams/{id}/replace` to `bitstream-page-routing.module.ts` that routes to `ReplaceBitstreamPageComponent`.
3. Created `SubmissionSectionUploadFileReplaceComponent` that provides a modal during item submission that provides the same functionality as `ReplaceBitstreamPageComponent`.
4. Added a replace button to `SubmissionSectionUploadFileComponent`, which opens the `SubmissionSectionUploadFileReplaceComponent` modal.
5. Updated `UploaderComponent` with a property `dragoverContainer` that defines the area in which dragOver events are processed. That allows dropping files into the submission modal without triggering the submission form itself.

How to test or review the PR:

**NOTE**: To properly test the feature, you need to run the REST changes mentioned in the References section above.

**Replace a bitstream through the item page.**
- Edit an item with a bitstream in the original bundle.
- In the bitstreams tab of the edit page, press the new replace button for a bitstream in the original bundle.
- Upload a new file to the bitstream replacement page and press save.
- Verify that the new file replaced the old one. 

**Replace a bitstream during submission.**
- Create a new item through the admin sidebar.
- Upload a file in the submission form.
- Scroll to the uploaded file section and press the new replace button
- Upload another file to the replace modal.
- Verify that the new file replaced the old one.

## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).

- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.

- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`

- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)

- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.

- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).

- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.

- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.

- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.

- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.

- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

Commit:
130076: Replace bitstreams from edit page

Commit:
130253: Add replace modal to item submission

Commit:
130253: Implement feedback 2025/04/24 (WIP)

Commit:
130253: Implement feedback 2025/04/24

Commit:
130478: Add replaceName param to file replace components

Commit:
130478: Implement feedback 2025/05/06

Commit:
Merge branch 'main' into w2p-130253_replace-bitstreams-from-submission

Commit:
131461: Limit submission replace button to new item versions

Commit:
131461: Implement feedback 2025-06-10

Commit:
Merge remote-tracking branch 'contributions/w2p-130253_replace-bitstreams-from-submission' into w2p-130253_replace-bitstreams-from-submission_contribute-main

Commit:
131461: Implement feedback 2025-06-10

Commit:
Merge branch 'main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

Commit:
136588: Use PUT and /content for replace endpoint

Commit:
Merge branch 'main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

# Conflicts:
#	src/app/submission/sections/upload/file/section-upload-file.component.spec.ts
#	src/app/submission/sections/upload/file/section-upload-file.component.ts

Commit:
Merge branch 'w2p-130253_replace-bitstreams-from-submission' into w2p-130253_replace-bitstreams-from-submission_contribute-main

# Conflicts:
#	src/app/bitstream-page/bitstream-page-routing.module.ts
#	src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.html
#	src/app/bitstream-page/replace-bitstream-page/replace-bitstream-page.component.ts
#	src/app/item-page/edit-item-page/item-bitstreams/item-edit-bitstream/item-edit-bitstream.component.html
#	src/app/submission/sections/upload/file/section-upload-file.component.html
#	src/app/submission/sections/upload/file/section-upload-file.component.spec.ts
#	src/app/submission/sections/upload/file/section-upload-file.component.ts

Commit:
Merge remote-tracking branch 'origin/main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

Commit:
Merge branch 'main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

# Conflicts:
#	src/assets/i18n/en.json5

Commit:
140086: Fix tests and formatting

Commit:
140086: Fix replace endpoint names

Commit:
140086: Add config support to disable bitstream replacement

Commit:
Merge branch 'main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

# Conflicts:
#	src/assets/i18n/en.json5

tdonohue reviewed (changes requested):
@ybnd and @gingyx : I reviewed & tested this with the backend.  Overall, the basics work, but I'm finding some buggy behavior in the UI.  I also have some inline code comments below.

Here's the bugs I've found.  In my testing, I'm always using the UI in production mode:
1. When replacing Bitstreams from "Bitstreams" tab, the newly replaced bitstream doesn't appear until you reload the page.
    * From an Item page, choose the "Administer" menu option
    * Go to the Bitstreams tab
    * Click the replace button next to a bitstream
    * Drag and drop a bitstream that has a **different** filename, and ensure "Use new filename" is turn on. Click Save
    * You'll return to the list of Bitstreams and the **old** bitstream name will still display.  Reload the page and you'll see the new bitstream
2. When using the new "Edit" form option (new in 10.0), the replace button doesn't always appear initially.  This is a bit harder to reproduce, but it happens most frequently when there is just one Bitstream
    * Find an Item with on bitstream
    * Select the "Edit" option on the Item page. You'll go to the new submission-like editing form.
    * On that form, see if the replace button appears next to the bitstream.  It often does not.  If it doesn't, then it will appear if you upload another file.
    * If you cannot reproduce this with the first item you try, then try another one. I'm finding that it loads for some items but not for others and I've yet to determine an exact pattern.
3. When creating a new version, the replace button doesn't always appear initially, especially if you save the submission for later.
    * Visit an item with a single bitstream
    * Select "Create a new version"
    * On the Submission form scroll down to where the Bitstreams are.  Check if the replace button exists (it often does not).  
    * If it does exist, then click "Save for later" on the submission form.  Go back and edit the submission, and often the replace button will not exist here either.
    * Overall, the replace button is very inconsistent in appearing. My opinion is that it should appear everywhere when enabled, even on a brand new submission.
 

Beyond those bugs, the functionality appears to work.  Again, I do also have some inline code suggestions below.

Commit:
Merge branch 'main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

Commit:
140390: Fix caching / stale issues with replacement button and make replacement button in submission forms again after new CRIS ports

atarix83 reviewed (changes requested):
Thanks @gingyx for this PR!

I’ve tested it and overall it looks good to me. I can also reproduce the same issue mentioned by @tdonohue here: https://github.com/DSpace/dspace-angular/pull/4368#pullrequestreview-4076991653.
I have just one minor request regarding the positioning of the buttons in the replace page/modal:
- In the replace page, the buttons are too close to each other and are aligned to the left, while in most of the application they are usually spaced and aligned to the right.
<img width="1142" height="571" alt="Image" src="https://github.com/user-attachments/assets/d9c42527-1cb6-4844-b028-43f27c07cde6" />

- The same layout issue appears in the modal.
<img width="1184" height="582" alt="Image" src="https://github.com/user-attachments/assets/b550e7f4-9e3e-424c-945d-d0d24a26df9d" />

It would be great to update the button positioning to keep the UI consistent across the application.

Commit:
140355: Add bitstream edit page display for related replacement versions

Commit:
140390: Add class level TSdoc

Commit:
Merge remote-tracking branch 'contributions/w2p-130253_replace-bitstreams-from-submission_contribute-main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

Commit:
140355: Move alerts to separate component and add test

ybnd committed:
Split canReplaceBitstreamFeature for administrative & submitter UI

ybnd committed:
Fix layout issues

These were typical Bootstrap 4 → 5 problems again

ybnd committed:
Fix cache invalidation issues pt. 2

Old Bitstream UUID is not guaranteed to exist
We also have to invalidate the parent Item in order to refresh derived Bundle lists

Drop /replace URL from history after successful upload (won't be accessible anymore)

ybnd commented:
Thanks again for the detailed feedback everyone!
We aimed to get everything addressed at once to avoid spamming the PRs again. In the most recent push you'll see
- A "linked Bitstreams" overview section on the edit Bitstream page
  There are a few caveats:
    - Because old Bitstreams are deleted on replacement, it's not possible to link to their edit page
    - It could be useful to consider adding a sort of "tombstone page" for deleted Bitstreams, but that would require considerable effort that's out of scope for these PRs
- Bitstreams can now be replaced on submission pages not tied to an in-progress version
    - Mentioned [[here](https://github.com/DSpace/DSpace/pull/10554#issuecomment-4180223603)](https://github.com/DSpace/DSpace/pull/10554#issuecomment-4180223603). This was confusing, especially after the recent PR that introduced submission-style "Edit" pages
- Resolved some cache invalidation issues
- Separate authorization checks for Bitstream replacements between the "Edit" and "Administer" pages (more info in the REST PR)
- Fixed layout issues

ybnd committed:
Fix stray lint issues

atarix83 reviewed (changes requested):
@gingyx @ybnd 
thanks for adressing the feedback. 
I only have one concern regarding a recent change: 
> Bitstreams can now be replaced on submission pages not tied to an in-progress version Mentioned [https://github.com/DSpace/DSpace/pull/10554#issuecomment-4180223603](Bitstream tracking for replacements and copies  DSpace#10554 (comment)). This was confusing, especially after the recent PR that introduced submission-style "Edit" pages

I don't know if it's done on purpose but, while I agree that replacing bitstreams for archived items in edit submission mode is correct, I don’t think this action should be available during a fresh submission (i.e. workspaceitem / workflowitem), unless the submission is explicitly related to the creation of a new version.
Allowing bitstream replacement in this context could also create confusion in the resulting item, as it may introduce replacement-related metadata that doesn’t really make sense for a first-time submission.

Commit:
140390: Don't show replace button in initial submission, use /edititems endpoint

ybnd commented:
@atarix83: @jensvannerum addressed your last point of feedback ↑

Note that it also requires this REST change: https://github.com/DSpace/DSpace/pull/10554/commits/2fb3b123a4839b28758a92cdf1f4f173906da26e

tdonohue reviewed (changes requested):
@ybnd : Thanks for the updates.  I've retested & re-reviewed today.  First off, all my prior code review feedback is addressed.  The bugs I reported previous are fixed, and good to see you found a way to stop the replace button from appearing on a new submission.

However, I'm now finding the Replace button no longer appears on the New Version form.  It appears we are still having difficulty getting the Replace button to appear there but *not* for a new submission.

To reproduce:
    * Ensure versioning is on and the replace function is enabled
    * Create a new Version for any Item
    * In the form that appears, scroll down and notice that the Replace button does **not** exist next to the bitstreams.

Beyond that bug, this frontend PR looks good to me.  All other feedback has been addressed & prior reported bugs have been fixed.

Commit:
140390: DO show replace button for new versions but not for (unversioned) new submission + cleanup code for nested subscriptions

jensvannerum commented:
@tdonohue I resolved this issue by checking if the current workspaceitem is versioned , if so we **do** show the button for workspace items. It's still never shown for workflow items and always shown for archived items (based on the back end `CanReplaceBitstreamSubmitter` authorization feature, that is).

tdonohue commented:
@jensvannerum : I've tested this today and I'm still seeing the same behavior as before.  The replace button does **not** appear if the Item is versioned.  However, now the replace button **will appear** if I upload a new file.  So, maybe it's a caching issue?

Here's what I'm doing:
1. I'm running the UI in production mode, with replacement enabled on the backend (`replace-bitstream.enabled = true`)
2. Login as an Admin
3. Find an Item with a few files already attached
4. Version that Item.  The Version form appears.  Notice the files do NOT have a replace button.
5. Drag and drop a new file to the Version form.  Now, the replace button appears next to each file.
6. Click "Save for Later" on this new version
7. Go to MyDSpace and find that in-progress new version. Edit it.  The files still do NOT have a replace button.
8. Drag and drop a new file into the form.  The replace button appears again next to each file.

So, the behavior seems *better*, but it's still not working properly to display the replace button right when the version form loads. 

(Sidenote: I have been able to verify though that a *new submission* never shows the replace button, which is correct behavior.  Even when you drag and drop new files to the submission, the replace button never appears.)

ybnd commented:
@tdonohue @jensvannerum looks like this only happens in production mode:

- When first opening the page the "replace" button doesn't appear
- As soon as you "touch" the file (e.g. set primary Bitstream on/off), the button appears

jensvannerum commented:
@tdonohue Thanks for catching this.

I was able to reproduce and fix it locally (amended to previous commit and force pushed). The issue was related to the item not and it's version being resolved correctly.

ybnd committed:
Merge remote-tracking branch 'origin/main' into w2p-130253_replace-bitstreams-from-submission_contribute-main

nwoodward commented:
> @jensvannerum : I've tested this today and I'm still seeing the same behavior as before. The replace button does **not** appear if the Item is versioned. However, now the replace button **will appear** if I upload a new file. So, maybe it's a caching issue?
> 
> Here's what I'm doing:
> 
> 1. I'm running the UI in production mode, with replacement enabled on the backend (`replace-bitstream.enabled = true`)
> 2. Login as an Admin
> 3. Find an Item with a few files already attached
> 4. Version that Item.  The Version form appears.  Notice the files do NOT have a replace button.
> 5. Drag and drop a new file to the Version form.  Now, the replace button appears next to each file.
> 6. Click "Save for Later" on this new version
> 7. Go to MyDSpace and find that in-progress new version. Edit it.  The files still do NOT have a replace button.
> 8. Drag and drop a new file into the form.  The replace button appears again next to each file.
> 
> So, the behavior seems _better_, but it's still not working properly to display the replace button right when the version form loads.
> 
> (Sidenote: I have been able to verify though that a _new submission_ never shows the replace button, which is correct behavior. Even when you drag and drop new files to the submission, the replace button never appears.)

I tested this workflow after the most recent changes, and now in step #4 the files **do** have a replace button. The replace buttons are still there after uploading a new file to the Version form. Likewise, the replace button is now there when I got to MyDSpace and click Edit.

tdonohue reviewed (changes requested):
@jensvannerum and @ybnd  : I've tested again today, and I've verified that the replace button is now appearing in all the proper locations (in production mode).

However, when I try to do a bitstream replacement, it only works when editing an archived Item.  It doesn't work during the versioning process.

Again, I'm testing in Production Mode.  Here's what I'm seeing:
1. Edit an Item via "Edit" form.  Replace a bitstream -> Works well.
2. Edit an Item via "Administer -> Bitstreams" tab. Replace a bitstream -> Also works well.
3. Version an Item. Replace a bitstream immediately -> 400 error occurs and bitstream is not replaced.
4. Version an Item. Save for later. Go back into MyDSpace and edit the new version. Replace a bitstream -> 400 error occurs and bitstream is not replaced.

The request that returns a 400 error looks like this:
```
POST http://localhost:8080/server/api/submission/workspaceitems/3411?replaceFile=08e6d62e-3003-4cc8-95ef-2ec521d78bb1&replaceName=true
```
When the 400 error occurs, I see the following in the backend logs :
```
2026-04-29 20:41:30,553 ERROR unknown abcc2615-d768-4018-ae60-a4f62b5f1f25 org.dspace.app.rest.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:400)
 org.dspace.app.rest.exception.DSpaceBadRequestException: For input string: "08e6d62e-3003-4cc8-95ef-2ec521d78bb1"
   at org.dspace.app.rest.submit.SubmissionService.findOriginalFile(SubmissionService.java:555)
   at org.dspace.app.rest.submit.SubmissionService.uploadFileToInprogressSubmission(SubmissionService.java:427)
   at org.dspace.app.rest.repository.WorkspaceItemRestRepository.upload(WorkspaceItemRestRepository.java:189)
   at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
   at java.base/java.lang.reflect.Method.invoke(Method.java:580)
   at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:360)
   at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
   at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
   at org.springframework.security.authorization.method.AuthorizationManagerBeforeMethodInterceptor.proceed(AuthorizationManagerBeforeMethodInterceptor.java:268)
   at org.springframework.security.authorization.method.AuthorizationManagerBeforeMethodInterceptor.attemptAuthorization(AuthorizationManagerBeforeMethodInterceptor.java:263)
   at org.springframework.security.authorization.method.AuthorizationManagerBeforeMethodInterceptor.invoke(AuthorizationManagerBeforeMethodInterceptor.java:196)
   at org.springframework.security.config.annotation.method.configuration.DeferringMethodInterceptor.invoke(DeferringMethodInterceptor.java:44)
   at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
   at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728)
   at org.dspace.app.rest.repository.WorkspaceItemRestRepository$$SpringCGLIB$$0.upload(<generated>)
   at org.dspace.app.rest.RestResourceController.uploadInternal(RestResourceController.java:628)
   at org.dspace.app.rest.RestResourceController.upload(RestResourceController.java:545)
...
Caused by: java.lang.NumberFormatException: For input string: "08e6d62e-3003-4cc8-95ef-2ec521d78bb1"
   at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
   at java.base/java.lang.Integer.parseInt(Integer.java:662)
   at java.base/java.lang.Integer.parseInt(Integer.java:778)
   at org.dspace.app.rest.submit.SubmissionService.findOriginalFile(SubmissionService.java:553)
```
While the error definitely is a backend error, I'm not sure if it's a bug on the backend or if the frontend is sending an unexpected request.

Other things I tested that were successful:
* I've verified that replacement works even when versioning is disabled.
* I've also re-verified that when replacement is disabled, none of the forms have the replacement button.

Finally, I agree with @nwoodward 's [note in the backend PR](https://github.com/DSpace/DSpace/pull/10554#issuecomment-4345384710) about providing a *warning* for replacement.  I think the Bitstream Replacement popup needs to have a display that is more similar to the Delete Bitstream popup:

<img width="501" height="195" alt="Image" src="https://github.com/user-attachments/assets/f23e994a-68cd-44dd-b7b9-979918d71b84" />

Because Replacement is a destructive action, the replacement popup should have the following:
* It should have text at the bottom that says something like "Once a file is replaced, this operation can't be undone."
* It MUST have a "Cancel" button.
* I'd also recommend it's header text "Replace [filename]" be in red, and the "Save" button be a red button that says "Replace File" (or similar) to stylize it similar to the delete popup above.

Beyond that, honestly, I think this feature is looking good overall.  It just obviously needs the remaining bugs fixed.

ybnd committed:
Telegraph that Bitstream replacement is destructive

- Style the submission/edit Bitstream replacement modal to appear more like the delete modal
- Make sure the "replace" button is disabled until we have a file
- Sync all changes to the Bitstream replacement page

ybnd commented:
@tdonohue 
Thanks for the feedback! 

The HTTP 400 issue was because the "edit" and "submission" endpoints treat the `replaceFile` parameter differently.
Adjusted the REST endpoint to accept UUIDs instead of indices to keep it simple.

I've adjusted the replace modal (submission, edit) as suggested, and mirrored the changes to the page (administer).
Small additional improvement: made the button unclickable until a file is uploaded.

tdonohue commented:
@ybnd : In case you hadn't noticed, this PR is failing a few lint checks after the last update. If you could quickly fix those that'd be great!

tdonohue reviewed (approved):
👍 Thanks @ybnd , @jensvannerum and @gingyx !  This PR now looks good to me.  I've retested today and verified that file replacement works properly both for existing items and during versioning.  Also verified new submissions work as normal (and no file replacement is available there).  Finally, I like the new look of the file replacement popup, which now looks like this:

<img width="1137" height="359" alt="Image" src="https://github.com/user-attachments/assets/ca2797a3-cfd4-4e48-bf56-063d5828288f" />

tdonohue commented:
@atarix83 and @nwoodward : Just a friendly reminder that these Bitstream Replacement PRs are at +1.  If either of you would like to help get this into 10.0, it just needs a second approval.

nwoodward reviewed (approved):
👍 Thanks @ConfusionOrb221 @gingyx @ybnd and @jensvannerum. I tested this again today using items and publications, with item-level versioning off and on, etc. and everything worked well. It adds useful functionality to DSpace, and I look forward to the follow-up PR in the future that will display old bitstream versions.
